README.rst
===================
django-calendardate
===================

*A calendar model with date metadata for querying against.*

Sometimes it is useful to have a range of dates and associated metadata stored in the database, rather than calculating on-the-fly, particularly when creating reports based on quarters, fiscal years, etc. This is where django-calendardate comes in. Also known as a date dimension table or calendar table, the inspiration for this project comes from `this article <https://www.sqlshack.com/designing-a-calendar-table/>`_.

django-calendardate provides:

* a model for storing metadata about a set of dates.
* a management command for generating the metadata


THIS PROJECT IS PRE_RELEASE. It works, but is missing documentation and tests. Be cautious.

Usage
-----

To process a range of dates (ignoring any duplicates):

    ``python manage.py process_dates 2021-01-01 2021-12-31``


To process a range of dates (overwriting any duplicates):

    ``python manage.py process_dates 2021-01-01 2021-12-31 --force``


Optional settings.py settings
-----------------------------

By default, django-calendardate assumes a fiscal year starts in October. If your fiscal year begins on a different month, set ``FISCAL_YEAR_START_MONTH`` to the month number (1-indexed). For instance January would be 1).

    ``FISCAL_YEAR_START_MONTH = 1``


Model Fields
------------

Each of the following are automatically calculated and inserted into the model using the ``process_dates`` management command.

    **calendar_date** (*DateField*)
        The actual date object

    **calendar_day** (*PositiveSmallIntegerField*)
        Number from 1 through 31

    **calendar_month** (*PositiveSmallIntegerField*)
        Month number from 1-12

    **calendar_year** (*PositiveSmallIntegerField*)
        Current year, eg: 2017, 2025, 1984

    **calendar_quarter** (*PositiveSmallIntegerField*)
        1-4, indicates quarter within the current year

    **fiscal_year** (*PositiveSmallIntegerField*)
        Current fiscal year, eg: 2017, 2025, 1984

    **fiscal_quarter** (*PositiveSmallIntegerField*)
        1-4, indicates fiscal quarter within the current fiscal year

    **day_of_week** (*PositiveSmallIntegerField*)
        Monday is 0 and Sunday is 6

    **day_of_isoweek** (*PositiveSmallIntegerField*)
        Monday is 1 and Sunday is 7

    **day_of_quarter** (*PositiveSmallIntegerField*)
        Number from 1-92, indicates the day # in the current quarter

    **day_of_year** (*PositiveSmallIntegerField*)
        Number from 1-366

    **week_of_month** (*PositiveSmallIntegerField*)
        Number from 1-6, indicates the number of week within the current month

    **week_of_year** (*PositiveSmallIntegerField*)
        Number from 1-53, indicates the number of week within the current year

    **isoweek_of_year** (*PositiveSmallIntegerField*)
        Number from 1-53, indicates the number of isoweek within the current year

    **is_weekday** (*BooleanField*)
        True if Monday-->Friday, False for Saturday/Sunday

    **is_leap_year** (*BooleanField*)
        True if current year is a leap year

    **days_in_month** (*PositiveSmallIntegerField*)
        Number of days in the current month

Model Properties
----------------

Each of the following properties are provided for each date.

    **get_day_name** (*property*)
        returns a string with the name of the day for the given date (e.g.: "Monday")

    **get_month_name** (*property*)
        returns a string with the name of the month for the given date (e.g.: "January")

    **get_month_abbreviated** (*property*)
        returns a string with the abbreviated name of the month for the given date (e.g.: "Jan")


Quick Example:
--------------

Say you have an Order model with a `order_date` field, and you want to query all of the orders that were placed in the third fiscal quarter of fiscal year 2021.

.. code-block:: python

    # Return list of  of dates in 3rd Qtr of FY21
    third_fiscal_qtr_dates = CalendarDate.objects.filter(fiscal_year=2021, fiscal_quarter=3).values_list('calendar_date', flat=True)
    
    # Filter on those dates
    third_qtr_orders = Order.objects.filter(order_date__in=third_fiscal_qtr_dates)


To Do
-----

1. Testing

2. Better documentation

3. Translations (strings already marked for translation)


CHANGELOG.md
Changelog
=========

## main

[View commits](https://github.com/OmenApps/django-calendardate/compare/0.1.1...HEAD)

## 0.1.1

* Updated README to provide some basic instructions and a bit of data about the CalendarDate model.


tests/urls.py
from django.conf import settings
from django.contrib import admin
from django.urls import path


app_name = "django_calendardate"

urlpatterns = [
    path(settings.ADMIN_URL, admin.site.urls),
]


tests/settings.py
import django

DEBUG = True
USE_TZ = True

SECRET_KEY = "nonsensevalue"

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.sqlite3",
        # "NAME": ":memory:",
        "NAME": "db",
    }
}

ADMIN_URL = "admin/"

ROOT_URLCONF = "tests.urls"

INSTALLED_APPS = [
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.messages",
    "django.contrib.sites",
    "django.contrib.sessions",
    "django.contrib.admin",
    "django_calendardate",
]

SITE_ID = 1


MIDDLEWARE = (
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
)

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.contrib.auth.context_processors.auth",
                "django.template.context_processors.debug",
                "django.template.context_processors.i18n",
                "django.template.context_processors.media",
                "django.template.context_processors.static",
                "django.template.context_processors.tz",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "handlers": {
        "stream": {
            "level": "DEBUG",
            "class": "logging.StreamHandler",
        },
    },
    "loggers": {
        "django.request": {
            "handlers": ["stream"],
            "level": "ERROR",
            "propagate": True,
        },
        "django_calendardate": {
            "handlers": ["stream"],
            "level": "ERROR",
            "propagate": True,
        },
    },
}

FISCAL_YEAR_START_MONTH = 10


django_calendardate/models.py
from django.db import models
from django.utils.translation import gettext as _


class CalendarDate(models.Model):
    calendar_date = models.DateField(_("Date"), unique=True, help_text=_("The actual date object"))
    calendar_day = models.PositiveSmallIntegerField(_("Calendar Day"), help_text=_("Number from 1 through 31"))
    calendar_month = models.PositiveSmallIntegerField(_("Calendar Month"), help_text=_("Month number from 1-12"))
    calendar_year = models.PositiveSmallIntegerField(
        _("Calendar Year"), help_text=_("Current year, eg: 2017, 2025, 1984")
    )
    calendar_quarter = models.PositiveSmallIntegerField(
        _("Calendar Quarter"), help_text=_("1-4, indicates quarter within the current year")
    )
    fiscal_year = models.PositiveSmallIntegerField(
        _("Fiscal Year"), help_text=_("Current fiscal year, eg: 2017, 2025, 1984")
    )
    fiscal_quarter = models.PositiveSmallIntegerField(
        _("Fiscal Quarter"), help_text=_("1-4, indicates fiscal quarter within the current fiscal year")
    )
    day_of_week = models.PositiveSmallIntegerField(_("Day of Week"), help_text=_("Monday is 0 and Sunday is 6"))
    day_of_isoweek = models.PositiveSmallIntegerField(_("Day of ISO Week"), help_text=_("Monday is 1 and Sunday is 7"))
    day_of_quarter = models.PositiveSmallIntegerField(
        _("Day in Quarter"), help_text=_("Number from 1-92, indicates the day # in the current quarter")
    )
    day_of_year = models.PositiveSmallIntegerField(_("Day in Year"), help_text=_("Number from 1-366"))
    week_of_month = models.PositiveSmallIntegerField(
        _("Week of Month"), help_text=_("Number from 1-6, indicates the number of week within the current month")
    )
    week_of_year = models.PositiveSmallIntegerField(
        _("Week of Year"), help_text=_("Number from 1-53, indicates the number of week within the current year")
    )
    isoweek_of_year = models.PositiveSmallIntegerField(
        _("ISO Week of Year"), help_text=_("Number from 1-53, indicates the number of isoweek within the current year")
    )
    is_weekday = models.BooleanField(
        _("Is Weekday"), default=False, help_text=_("True if Monday-->Friday, False for Saturday/Sunday")
    )
    is_leap_year = models.BooleanField(
        _("Is Leap Year"), default=False, help_text=_("True if current year is a leap year")
    )
    days_in_month = models.PositiveSmallIntegerField(
        _("Days in current month"), help_text=_("Number of days in the current month")
    )

    class Meta:
        verbose_name = _("Calendar Date")
        verbose_name_plural = _("Calendar Dates")

        get_latest_by = ["-calendar_date"]
        ordering = ["-calendar_date"]

        indexes = [
            models.Index(fields=["calendar_date"]),
            models.Index(fields=["calendar_year", "calendar_month"]),
            models.Index(fields=["calendar_year", "calendar_quarter"]),
            models.Index(fields=["calendar_year", "week_of_year"]),
            models.Index(fields=["fiscal_year", "fiscal_quarter"]),
            models.Index(fields=["is_weekday"]),
        ]

    def __str__(self):
        return f"{self.calendar_date}"

    @property
    def get_day_name(self):
        day_names = {
            1: _("Monday"),
            2: _("Tuesday"),
            3: _("Wednesday"),
            4: _("Thursday"),
            5: _("Friday"),
            6: _("Saturday"),
            7: _("Sunday"),
        }
        return day_names[self.day_of_isoweek]

    @property
    def get_month_name(self):
        month_names = {
            1: _("January"),
            2: _("February"),
            3: _("March"),
            4: _("April"),
            5: _("May"),
            6: _("June"),
            7: _("July"),
            8: _("August"),
            9: _("September"),
            10: _("October"),
            11: _("November"),
            12: _("December"),
        }
        return month_names[self.calendar_month]

    @property
    def get_month_abbreviated(self):
        month_names = {
            1: _("Jan"),
            2: _("Feb"),
            3: _("Mar"),
            4: _("Apr"),
            5: _("May"),
            6: _("Jun"),
            7: _("Jul"),
            8: _("Aug"),
            9: _("Sep"),
            10: _("Oct"),
            11: _("Nov"),
            12: _("Dec"),
        }
        return month_names[self.calendar_month]


django_calendardate/date_utilities.py
import bisect
import calendar
import datetime
import logging
import math

from django.conf import settings

logger = logging.getLogger("django_calendardate")


fy_start_month = getattr(settings, "FISCAL_YEAR_START_MONTH", 10)


def process_date(date):
    """
    Returns a dictionary with all of the below processes completed,
    which can then be imported directly to CalendarDate model.
    """

    metadata = {}
    metadata["calendar_date"] = date
    metadata["calendar_day"] = get_calendar_day(date)
    metadata["calendar_month"] = get_calendar_month(date)
    metadata["calendar_year"] = get_calendar_year(date)
    metadata["calendar_quarter"] = get_calendar_quarter(date)
    metadata["fiscal_year"] = get_fiscal_year(date)
    metadata["fiscal_quarter"] = get_fiscal_quarter(date)
    metadata["day_of_week"] = get_day_of_week(date)
    metadata["day_of_isoweek"] = get_day_of_isoweek(date)
    metadata["day_of_quarter"] = get_day_of_quarter(date)
    metadata["day_of_year"] = get_day_of_year(date)
    metadata["week_of_month"] = get_week_of_month(date)
    metadata["week_of_year"] = get_week_of_year(date)
    metadata["isoweek_of_year"] = get_isoweek_of_year(date)
    metadata["is_weekday"] = get_is_weekday(date)
    metadata["is_leap_year"] = get_is_leap_year(date)
    metadata["days_in_month"] = get_days_in_month(date)

    logger.debug(f"process_date metadata: {metadata}")

    return metadata


def get_quarter_begin(date):
    quarter_begins = [datetime.date(date.year, month, 1) for month in (1, 4, 7, 10)]
    idx = bisect.bisect(quarter_begins, date)
    return quarter_begins[idx - 1]


def get_calendar_day(date):
    return date.day


def get_calendar_month(date):
    return date.month


def get_calendar_year(date):
    return date.year


def get_calendar_quarter(date):
    return int(math.ceil(date.month / 3))


def get_fiscal_year(date):
    return date.year if date.month < fy_start_month or fy_start_month == 1 else date.year + 1


def get_fiscal_quarter(date):

    if fy_start_month in [1, 4, 7, 10]:
        quarter_starts = [1, 4, 7, 10]
        quarter_idx = quarter_starts.index(fy_start_month)
    elif fy_start_month in [2, 5, 8, 11]:
        quarter_starts = [2, 5, 8, 11]
        quarter_idx = quarter_starts.index(fy_start_month)
    else:
        quarter_starts = [3, 6, 9, 12]
        quarter_idx = quarter_starts.index(fy_start_month)

    quarter_start_dates = [datetime.date(date.year, month, 1) for month in quarter_starts]

    quarter = bisect.bisect(quarter_start_dates, date) - quarter_idx
    return quarter if quarter > 0 else quarter + 4


def get_day_of_week(date):
    return date.weekday()


def get_day_of_isoweek(date):
    return date.isoweekday()


def get_day_of_quarter(date):
    delta = date - get_quarter_begin(date) + datetime.timedelta(days=1)
    return delta.days


def get_day_of_year(date):
    return int(date.strftime("%j"))


def get_week_of_month(date):
    def week_of_month(date_object):
        """Returns the week of the month for the specified date."""
        first_day = date_object.replace(day=1)
        dom = date_object.day
        adjusted_dom = dom + first_day.weekday()
        return int(math.ceil(adjusted_dom / 7.0))

    return week_of_month(date)


def get_week_of_year(date):
    return int(date.strftime("%U")) + 1


def get_isoweek_of_year(date):
    return int(date.strftime("%W")) + 1


def get_is_weekday(date):
    return True if date.weekday() < 5 else False


def get_is_leap_year(date):
    year = date.year
    return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)


def get_days_in_month(date):
    _, days = calendar.monthrange(date.year, date.month)
    return days


django_calendardate/management/commands/process_dates.py
"""
Based on the date range provided, this command adds CalendarDate
objects for any non-yet-existing dates in that range.
"""
import logging

from datetime import datetime, timedelta
from django.core.management.base import BaseCommand, CommandError
from ...models import CalendarDate
from ...date_utilities import process_date

logger = logging.getLogger("django_calendardate")


class Command(BaseCommand):
    help = """Processes (and creates) dates for the CalendarDate model.
        Use: `manage.py process_dates 2000-01-01 2005-12-31`

        Optionally, the --force argument will overwrite existing dates.
        """

    def add_arguments(self, parser):
        parser.add_argument("start_date", type=str, help="Start of the date period, formatted as: YYYY-MM-DD")
        parser.add_argument("end_date", type=str, help="End of the date period, formatted as: YYYY-MM-DD")

        parser.add_argument(
            "--force",
            action="store_true",
            help="Force overwrite of existing dates",
        )

    def handle(self, *args, **options):
        start_date = datetime.strptime(options["start_date"], "%Y-%m-%d").date()
        end_date = datetime.strptime(options["end_date"], "%Y-%m-%d").date()
        delta = timedelta(days=1)

        logger.debug(f"Processing dates from {start_date} to {end_date}")

        while start_date <= end_date:

            date_metadata = process_date(start_date)

            obj, created = CalendarDate.objects.update_or_create(**date_metadata)

            start_date += delta

            if created:
                self.stdout.write(
                    self.style.SUCCESS(f"Successfully created/verified CalendarDate for {obj.calendar_date}.")
                )
            else:
                if options["force"]:
                    self.stdout.write(self.style.SUCCESS(f"CalendarDate for {obj.calendar_date} updated."))
                else:
                    self.stdout.write(
                        self.style.SUCCESS(f"CalendarDate for {obj.calendar_date} already exists. Skipping.")
                    )


django_calendardate/migrations/0001_initial.py
# Generated by Django 3.1.7 on 2021-02-21 17:10

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = []

    operations = [
        migrations.CreateModel(
            name="CalendarDate",
            fields=[
                ("id", models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID")),
                (
                    "calendar_date",
                    models.DateField(help_text="The actual date object", unique=True, verbose_name="Date"),
                ),
                (
                    "calendar_day",
                    models.PositiveSmallIntegerField(help_text="Number from 1 through 31", verbose_name="Calendar Day"),
                ),
                (
                    "calendar_month",
                    models.PositiveSmallIntegerField(help_text="Month number from 1-12", verbose_name="Calendar Month"),
                ),
                (
                    "calendar_year",
                    models.PositiveSmallIntegerField(
                        help_text="Current year, eg: 2017, 2025, 1984", verbose_name="Calendar Year"
                    ),
                ),
                (
                    "calendar_quarter",
                    models.PositiveSmallIntegerField(
                        help_text="1-4, indicates quarter within the current year", verbose_name="Calendar Quarter"
                    ),
                ),
                (
                    "fiscal_year",
                    models.PositiveSmallIntegerField(
                        help_text="Current fiscal year, eg: 2017, 2025, 1984", verbose_name="Fiscal Year"
                    ),
                ),
                (
                    "fiscal_quarter",
                    models.PositiveSmallIntegerField(
                        help_text="1-4, indicates fiscal quarter within the current fiscal year",
                        verbose_name="Fiscal Quarter",
                    ),
                ),
                (
                    "day_of_week",
                    models.PositiveSmallIntegerField(
                        help_text="Monday is 0 and Sunday is 6", verbose_name="Day of Week"
                    ),
                ),
                (
                    "day_of_isoweek",
                    models.PositiveSmallIntegerField(
                        help_text="Monday is 1 and Sunday is 7", verbose_name="Day of ISO Week"
                    ),
                ),
                (
                    "day_of_quarter",
                    models.PositiveSmallIntegerField(
                        help_text="Number from 1-92, indicates the day # in the current quarter",
                        verbose_name="Day in Quarter",
                    ),
                ),
                (
                    "day_of_year",
                    models.PositiveSmallIntegerField(help_text="Number from 1-366", verbose_name="Day in Year"),
                ),
                (
                    "week_of_month",
                    models.PositiveSmallIntegerField(
                        help_text="Number from 1-6, indicates the number of week within the current month",
                        verbose_name="Week of Month",
                    ),
                ),
                (
                    "week_of_year",
                    models.PositiveSmallIntegerField(
                        help_text="Number from 1-53, indicates the number of week within the current year",
                        verbose_name="Week of Year",
                    ),
                ),
                (
                    "isoweek_of_year",
                    models.PositiveSmallIntegerField(
                        help_text="Number from 1-53, indicates the number of isoweek within the current year",
                        verbose_name="ISO Week of Year",
                    ),
                ),
                (
                    "is_weekday",
                    models.BooleanField(
                        default=False,
                        help_text="True if Monday-->Friday, False for Saturday/Sunday",
                        verbose_name="Is Weekday",
                    ),
                ),
                (
                    "is_leap_year",
                    models.BooleanField(
                        default=False, help_text="True if current year is a leap year", verbose_name="Is Leap Year"
                    ),
                ),
                (
                    "days_in_month",
                    models.PositiveSmallIntegerField(
                        help_text="Number of days in the current month", verbose_name="Days in current month"
                    ),
                ),
            ],
            options={
                "verbose_name": "Calendar Date",
                "verbose_name_plural": "Calendar Dates",
                "ordering": ["-calendar_date"],
                "get_latest_by": ["-calendar_date"],
            },
        ),
        migrations.AddIndex(
            model_name="calendardate",
            index=models.Index(fields=["calendar_date"], name="django_cale_calenda_95378f_idx"),
        ),
        migrations.AddIndex(
            model_name="calendardate",
            index=models.Index(fields=["calendar_year", "calendar_month"], name="django_cale_calenda_fcdc17_idx"),
        ),
        migrations.AddIndex(
            model_name="calendardate",
            index=models.Index(fields=["calendar_year", "calendar_quarter"], name="django_cale_calenda_b87451_idx"),
        ),
        migrations.AddIndex(
            model_name="calendardate",
            index=models.Index(fields=["calendar_year", "week_of_year"], name="django_cale_calenda_a55a54_idx"),
        ),
        migrations.AddIndex(
            model_name="calendardate",
            index=models.Index(fields=["fiscal_year", "fiscal_quarter"], name="django_cale_fiscal__592cc1_idx"),
        ),
        migrations.AddIndex(
            model_name="calendardate",
            index=models.Index(fields=["is_weekday"], name="django_cale_is_week_c191dd_idx"),
        ),
    ]
